### PR TITLE
Add documentation for torch.overrides submodule.

### LINKED ISSUE #48087: Add documentation for the torch.overrides submodule.
## 📚 Documentation
As mentioned in https://github.com/pytorch/pytorch/issues/47697#issuecomment-725743901, the submodule `torch.overrides` was made public but has no user-facing documentation.

There are two separate tasks here:

- Add the documentation of the submodule.
- Expand the coverage of the docstrings so they are more descriptive, include examples, and are user friendly.

cc @hameerabbasi @rgommers @jlin27 @mruberry

### PR DESCRIPTION
**Pull request by hameerabbasi**
Fixes #48087

### CODE REVIEW

**Comment by hameerabbasi** [comment]
@ezyang Now that these things are being documented, it's our last chance to make anything private. So let me know if you'd make anything else private.

File: torch/overrides.py [hunk 1]
```diff
@@ -28,15 +30,33 @@
 import torch
 from torch._C import _is_torch_function_enabled, _disabled_torch_function_impl
 
+__all__ = [
+    "get_ignored_functions",
+    "get_testing_overrides",
```
Comment (hameerabbasi) [inline torch/overrides.py:35]: Related to https://github.com/pytorch/pytorch/pull/48170#issuecomment-729615574, I'd make these two private as well.

Comment (ezyang) [reply torch/overrides.py:35]: get_ignored_functions and get_testing_overrides? I agree they should be private too.

**Comment by ezyang** [comment]
I looked over all the funcs, and I agree with your assessment of what should be public and private.

File: torch/overrides.py [hunk 2]
```diff
@@ -969,6 +998,34 @@ def get_testing_overrides() -> Dict[Callable, Callable]:
     ret.update(ret2)
     return ret
 
+def wrap_torch_function(dispatcher: Callable):
+    """Wraps a given function with ``__torch_function__`` -related functionality.
```
Comment (ezyang) [inline torch/overrides.py:1002]: Probably good to tell people not to use this if they care about performance

File: torch/overrides.py [hunk 3]
```diff
@@ -1055,13 +1110,13 @@ def handle_torch_function(
...
     -------
-    Result from calling `implementation()` or an `__torch_function__`
-    method, as appropriate.
+    object
+        Result from calling ``implementation`` or an ``__torch_function__``
+        method, as appropriate.
 
     Raises
     ------
     TypeError : if no implementation is found.
```
Comment (ezyang) [inline torch/overrides.py:1119]: I think the little sample code for all the places in our codebase where we call it manually would be very instructive here.

File: docs/source/notes/extending.rst [hunk 4]
```diff
@@ -605,14 +607,14 @@ provides a developer-facing API for ensuring full support for
 changes without warning in the future.
 
 First, to get a listing of all overridable functions, use
-``torch.overrides.get_overridable_functions``. This returns a dictionary whose
+``torch.overrides._get_overridable_functions``. This returns a dictionary whose
```
Comment (ezyang) [inline docs/source/notes/extending.rst:610]: Ooh... I forgot that you refer to this explicitly in the docs. Referencing a private function here is kind of weird. Perhaps that means this particular function should be kept public. `_get_testing_overrides` and `_get_ignored_functions` are the ones that definitely should not be public.

Comment (hameerabbasi) [reply docs/source/notes/extending.rst:610]: Version 1.7 is the only version where these were public. Before that, the submodule was private. v1.7.1 is the only version where the docs might show it as public.

Comment (ezyang) [reply docs/source/notes/extending.rst:610]: Sorry, it doesn't have to do with the historical context is. What I mean is, it's weird to mark a function with a leading underscore, and then immediately turn around and in the docs say, "Hey, you should call this private function to find out what functions to implement it with!" It's like saying, "This thing, which you're not supposed to use, is so useful that we are going to tell you about it in the public documentation hint hint wink wink"

Comment (hameerabbasi) [reply docs/source/notes/extending.rst:610]: I'll make these public.

**Comment by hameerabbasi** [comment]
The CI failures are all unrelated.

* For the `flake8` failure, see #24188.
* CircleCI macOS build had an infrastructure issue.
* The failing test for ROCm is `test_streaming_backward_sync_graph_root (__main__.TestCuda)` (`Tensors failed to compare as equal!`), which has nothing to do with this PR either.

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**